cps/kni.c

```diff
@@ -613,6 +587,152 @@ mnl_socket_recvfrom_flags(const struct mnl_socket *nl, void *buf, size_t bufsiz,
 	return ret;
 }
 
+static int
+new_route(struct route_update *update, struct gk_config *gk_conf)
+{
+	int ret;
+	uint16_t proto;
+	char ip_buf[INET6_ADDRSTRLEN];
+	char gw_buf[INET6_ADDRSTRLEN];
+	char ipp_buf[INET6_ADDRSTRLEN + 4];
+	char gt_buf[INET6_ADDRSTRLEN];
+	int prefix_fib_id, gw_fib_id;
+	struct ipaddr *gt_addr = NULL;
+	struct gk_fib *prefix_fib = NULL;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (update->family == AF_INET6) {
+		proto = ETHER_TYPE_IPv6;
+
+		prefix_fib_id = lpm_lookup_ipv6(
+			ltbl->lpm6, update->ip.v6.s6_addr);
+		if (prefix_fib_id < 0) {
+			gw_fib_id = lpm_lookup_ipv6(
+				ltbl->lpm6, update->gw.v6.s6_addr);
+			if (gw_fib_id < 0)
+				return -1;
+			gw_fib = &ltbl->fib_tbl6[gw_fib_id];
+		} else
+			prefix_fib = &ltbl->fib_tbl6[prefix_fib_id];
+
+		if (inet_ntop(AF_INET6, &update->ip.v6.s6_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+
+		if (inet_ntop(AF_INET6, &update->gw.v6.s6_addr,
+				gw_buf, sizeof(gw_buf)) == NULL)
+			return -1;
+	} else {
+		proto = ETHER_TYPE_IPv4;
+
+		prefix_fib_id = lpm_lookup_ipv4(ltbl->lpm, update->ip.v4.s_addr);
+		if (prefix_fib_id < 0) {
+			gw_fib_id = lpm_lookup_ipv4(
+				ltbl->lpm, update->gw.v4.s_addr);
+			if (gw_fib_id < 0)
+				return -1;
+			gw_fib = &ltbl->fib_tbl[gw_fib_id];
+		} else
+			prefix_fib = &ltbl->fib_tbl[prefix_fib_id];
+
+		if (inet_ntop(AF_INET, &update->ip.v4.s_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+
+		if (inet_ntop(AF_INET, &update->gw.v4.s_addr,
+				gw_buf, sizeof(gw_buf)) == NULL)
+			return -1;
+	}
+
+	if (prefix_fib && prefix_fib->action == GK_FWD_GRANTOR &&
+			prefix_fib->u.grantor.gt_addr.proto == proto)
+		gt_addr = &prefix_fib->u.grantor.gt_addr;
+	else if (prefix_fib && prefix_fib->action == GK_GET_GRANTOR &&
+			prefix_fib->u.grantor_ip.proto == proto)
+		gt_addr = &prefix_fib->u.grantor_ip;
+
+	if (gt_addr != NULL) {
+		if (proto == ETHER_TYPE_IPv4) {
+			if (inet_ntop(AF_INET, &gt_addr->ip.v4.s_addr,
+					gt_buf, sizeof(gt_buf)) == NULL)
+				return -1;
+		} else {
+			if (inet_ntop(AF_INET, &gt_addr->ip.v6.s6_addr,
+					gt_buf, sizeof(gt_buf)) == NULL)
+				return -1;
+		}
+
+		return add_fib_entry(ipp_buf, gt_buf, gw_buf,
```

Comment:
It seems like this function is converting the numerical representations of the prefix and IP addresses to the string representations only because `add_fib_entry()` accepts string parameters, but then it converts them back to numerical form in that function anyway.

Is there a good solution where we can avoid doing this conversion back and forth? For example, perhaps `add_fib_entry()` can be changed to factor out the code that converts the strings to the numerical form.

Reply:
Yes, since we allow the administrators to configure the LPM table in Lua scripts, which will accept the configurations as a string. However, here it accepts the configuration from CPS, which has the numerical representations of the prefix and IP addresses. If the function accepts the numerical representation, then the administrators may have to do the conversion. Probably, we can have two versions of the function ``add_fib_entry()``?